tests/unit/test_charm.py

```diff
@@ -40,47 +38,43 @@
 CREATE_CLUSTER_CONF_PATH = "/etc/postgresql-common/createcluster.d/pgcharm.conf"
 
 
-# @pytest.mark.usefixtures("juju_has_secrets")
-class TestCharm(unittest.TestCase):
-    def setUp(self):
-        self._peer_relation = PEER
-        self._postgresql_container = "postgresql"
-
-        self.harness = Harness(PostgresqlOperatorCharm)
-        self.addCleanup(self.harness.cleanup)
-        self.harness.begin()
-        self.charm = self.harness.charm
-        self.rel_id = self.harness.add_relation(self._peer_relation, self.charm.app.name)
-        self.harness.add_relation("upgrade", self.charm.app.name)
-
-    @pytest.fixture
-    def use_caplog(self, caplog):
-        self._caplog = caplog
-
-    @patch_network_get(private_address="1.1.1.1")
-    @patch("charm.subprocess.check_call")
-    @patch("charm.snap.SnapCache")
-    @patch("charm.PostgresqlOperatorCharm._install_snap_packages")
-    @patch("charm.PostgresqlOperatorCharm._reboot_on_detached_storage")
-    @patch(
+@pytest.fixture(autouse=True)
+def harness():
+    harness = Harness(PostgresqlOperatorCharm)
+    harness.begin()
+    harness.add_relation("upgrade", harness.charm.app.name)
+    yield harness
+    harness.cleanup()
+
+
+# This causes every test defined in this file to run 2 times, each with
+# charm.JujuVersion.has_secrets set as True or as False
+@pytest.fixture(params=[True, False], autouse=True)
+def _has_secrets(request, monkeypatch):
+    monkeypatch.setattr("charm.JujuVersion.has_secrets", PropertyMock(return_value=request.param))
+    return request.param
+
+
+@patch_network_get(private_address="1.1.1.1")
+def test_on_install(harness):
+    with patch("charm.subprocess.check_call") as _check_call, patch(
+        "charm.snap.SnapCache"
+    ) as _snap_cache, patch(
+        "charm.PostgresqlOperatorCharm._install_snap_packages"
+    ) as _install_snap_packages, patch(
+        "charm.PostgresqlOperatorCharm._reboot_on_detached_storage"
+    ) as _reboot_on_detached_storage, patch(
         "charm.PostgresqlOperatorCharm._is_storage_attached",
         side_effect=[False, True, True],
-    )
-    def test_on_install(
-        self,
-        _is_storage_attached,
-        _reboot_on_detached_storage,
-        _install_snap_packages,
-        _snap_cache,
-        _check_call,
-    ):
+    ) as _is_storage_attached:
+        harness.add_relation(PEER, harness.charm.app.name)
```

Comment:
I think it may be possible to move this to the fixture (to avoid forgetting about it when creating new tests and also have it in all tests, as we should have a peer relation present in pretty much every hook execution). For the calls that need the relation ID (`rel_id = harness.add_relation(PEER, harness.charm.app.name)`), you probably may use `harness.model.get_relation(PEER).id`).